File: 0653_two_sum_IV/find_target.py
def find_target(root, k):
    """
    Time: O(n)
    Space: O(n)
    """
    if not root:
        return False
    return find(root, set(), k)

def find(root, seen, k):
    if not root:
        return False
    compliment = k - root.val
    if compliment in seen:
        return True
    seen.add(root.val)
    return find(root.left, seen, k) or find(root.right, seen, k)

File: 0653_two_sum_IV/find_target.java
public class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode() {}
    TreeNode(int val) { this.val = val; }
    TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}

class Solution {
    // Time: O(n)
    // Space: O(n)
    public boolean findTarget(TreeNode root, int k) {
        return dfs(root, new HashSet<>(), k);
    }
    
    public boolean dfs(TreeNode root, HashSet<Integer> set, int k) {
        if (root == null) {
            return false;
        }
        int compliment = k - root.val;
        if (set.contains(compliment)) {
            return true;
        }
        set.add(root.val);
        return dfs(root.left, set, k) || dfs(root.right, set, k);
    }
}